src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2701,8 +2701,17 @@ BuildRoutesForInsert(Query *query, DeferredErrorMessage **planningError)
 	{
 		InsertValues *insertValues = (InsertValues *) lfirst(insertValuesCell);
 		List *prunedShardIntervalList = NIL;
-		Expr *partitionValueExpr = (Expr *) strip_implicit_coercions(
-			(Node *) insertValues->partitionValueExpr);
+		Node *partitionValueExpr = (Node *) insertValues->partitionValueExpr;
+		partitionValueExpr = strip_implicit_coercions(partitionValueExpr);
+
+		/*
+		 * By evaluating constant expressions an expression such as 2 + 4
+		 * will become const 6. That way we can use them as a partition column
+		 * value. Normally the planner evaluates constant expressions, but we
+		 * may be working on the original query tree here. So we do it here
+		 * explicitely before checking that the partition value is a const.
+		 */
+		partitionValueExpr = eval_const_expressions(NULL, partitionValueExpr);
```

Comment:
In Citus code, we mostly rely on `eval_const_expression`. But, I was considering to use [`expression_planner`](https://github.com/postgres/postgres/blob/92c58fd94801dd5c81ee20e26c5bb71ad64552a8/src/backend/optimizer/plan/planner.c#L6128) at some point. 

Do you see any difference for this case? I have no  idea, just bringing up as I was considering to use `expression_planner` over `eval_const_expressions ` if we think it is better at some point

Reply:
I took a look at the source code of `expression_planner`. The only difference is it also calls `fix_opfuncids` after it called `eval_const_expressions`. This doesn't matter in either of the calls in this PR:
1. `fix_opfuncids` doesn't do anything for `T_Const`, so we don't care about it at this callsite of `eval_const_expressions`. Because if we don't get a `T_Const` we error out.
2. In the other callsite, `fix_opfuncids` is already called in `citus_evaluate_expr`. So we would do it twice.

Since `fix_opfuncids` is another `expression_tree_walker`, I think it's better to leave them as is to not cause unnecessary slowdowns.

I'll add this information to the respective comments.